Prevent duplicate failed order emails

annemirasol opened the pull request:
When an order changes status from pending/on-hold to failed, [a failed-order email is triggered in core](https://github.com/woocommerce/woocommerce/blob/0f2fa403f1ad0dc26bdd82ec3e901cb1794201e6/plugins/woocommerce/includes/emails/class-wc-email-failed-order.php#L40). In our Stripe webhooks code, we also manually trigger these emails when certain processes fail, e.g. `payment_intent.payment_failed`. This causes duplicate emails for some scenarios.

<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #1229

## Changes proposed in this Pull Request:
Before manually triggering a failed-order email, we first check if the status change to `failed` will not trigger its own email send.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions
1. Stripe webhooks need to be set up and working:
    - Use a tunneling service like ngrok or Jurassic Tube, -or-
    - Test these changes on publicly accessible servers
2. Install the WP Mail Logging plugin: https://wordpress.org/plugins/wp-mail-logging/
4. As a shopper, add a product to cart and go to checkout.
5. Use `4000 0084 0000 1629` as the credit card number, with any expiry date and CVC.
    - This should trigger 3D Secure 2 authentication, but payment will eventually fail. 
    - Click "Complete" to authenticate. 
    - You should see the error banner above the payment element, with a message like `Your card has been declined.`
6. Go to the WP Mail Logging dashboard: `wp-admin/admin.php?page=wpml_plugin_log`
7. You should see a failed order email, with the subject `Order #XY has failed`.
    - You should see only one failed order email.
    - Note: customers re-attempting and again failing will trigger another failed order email. This is expected.


<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

Commit:
Prevent duplicate failed order emails

When an order changes status from pending/on-hold to failed,
a failed-order email is triggered by core. In our webhooks code,
we also manually trigger these emails when certain Stripe processes
fail. This causes duplicate emails for some scenarios.

Commit:
Merge branch 'develop' into fix/1229-duplicate-failed-order-emails

Commit:
Use strict comparisons

Commit:
Add changelog and readme entries

Commit:
Remove error logging

Commit:
Merge branch 'develop' into fix/1229-duplicate-failed-order-emails

wjrosa commented on includes/class-wc-stripe-webhook-handler.php:
```diff
@@ -1115,6 +1127,31 @@ protected function handle_deferred_payment_intent_succeeded( $order, $intent_id
+	 * @param string $status The status of the order prior to setting it to failed.
+	 * @return bool
+	 */
+	private function has_failed_order_email_hooks( $status ) {
```
Smart solution 👍

wjrosa reviewed (changes requested):
Unfortunately, it is not working as expected for me (still receiving 2 emails). I have included the following line to log your new method after `$send_email = ! $this->has_failed_order_email_hooks( $order->get_status() );`:
```
					WC_Stripe_Logger::error( 'has_failed_order_email_hooks: ' . wc_bool_to_string( ! $send_email ) . ' / ' . $order->get_status() );
```

For me, on a JN site, `has_failed_order_email_hooks` always return `false` and `$order->get_status()` is always `pending` at that point.

Commit:
Fix typo on email class name

Commit:
Move logic to implementing method

The Stripe webhook code should not be too coupled with email implem
details, so we are moving the dedup logic to the method that actually
handles the email sending.

Commit:
Merge branch 'develop' into fix/1229-duplicate-failed-order-emails

annemirasol commented:
@wjrosa So sorry about that! I think during testing, I wasn't properly ensuring I was failing on a fresh order ID, so it was only sending one email because there was no status update actually happening! 🫠 

Anyway, there were a few things wrong in my old code:
- typo on the class name ☠️ 
- I was using the class name in the callback param, when I should be using a class instance

Fixes:
- actually use the correct class and instance
- move the logic to inside `WC_Stripe_Payment_Gateway`'s `send_failed_order_email()` -- this one is more of a rework than a fix. I realized the Stripe webhook code should not know too much about how emails are sent, and so I moved the dedupe logic to the method that is actually in charge of sending emails. Let me know what you think of this!

Tested, making sure orders are fresh!

wjrosa commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -622,13 +622,33 @@ public function process_response( $response, $order ) {
 	 * @param int $order_id
+	 * @param array $status_update Optional. Use fields 'to' and 'from' to provide context
+	 *    when a status change has occurred for the order.
 	 * @return null
```
```suggestion
	 * @return void
```

wjrosa commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -622,13 +622,33 @@ public function process_response( $response, $order ) {
+		// will trigger its own failed order email.
+		if (
+			isset( $status_update['to'] ) &&
+			isset( $status_update['from'] ) &&
```
You can replace both `isset` with `isset( $status_update['to'], $status_update['from'] )`

annemirasol replied:
Thanks, TIL!

wjrosa reviewed (approved):
Thanks for fixing the issues, Anne! It now works as expected 👍 . Left some minor (non-blocker) comments.

Commit:
Fix  docblock return param

Co-authored-by: Wesley Rosa <wesleyjrosa@gmail.com>

Commit:
Merge branch 'develop' into fix/1229-duplicate-failed-order-emails

Commit:
Refactor: use multiple params for isset()
